solr/core/src/java/org/apache/solr/servlet/QueryRateLimiter.java

```diff
@@ -17,39 +17,82 @@
+import org.apache.zookeeper.data.Stat;
 
-import static org.apache.solr.servlet.RateLimitManager.DEFAULT_CONCURRENT_REQUESTS;
-import static org.apache.solr.servlet.RateLimitManager.DEFAULT_SLOT_ACQUISITION_TIMEOUT_MS;
+import static org.apache.solr.servlet.RateLimiterConfig.RL_CONFIG_KEY;
 
 /** Implementation of RequestRateLimiter specific to query request types. Most of the actual work is delegated
  *  to the parent class but specific configurations and parsing are handled by this class.
  */
 public class QueryRateLimiter extends RequestRateLimiter {
-  final static String IS_QUERY_RATE_LIMITER_ENABLED = "isQueryRateLimiterEnabled";
-  final static String MAX_QUERY_REQUESTS = "maxQueryRequests";
-  final static String QUERY_WAIT_FOR_SLOT_ALLOCATION_INMS = "queryWaitForSlotAllocationInMS";
-  final static String QUERY_GUARANTEED_SLOTS = "queryGuaranteedSlots";
-  final static String QUERY_ALLOW_SLOT_BORROWING = "queryAllowSlotBorrowing";
-
-  public QueryRateLimiter(FilterConfig filterConfig) {
-    super(constructQueryRateLimiterConfig(filterConfig));
+
+  public QueryRateLimiter(SolrZkClient solrZkClient) {
+    super(constructQueryRateLimiterConfig(solrZkClient));
+  }
+
+  @SuppressWarnings({"unchecked"})
+  public void processConfigChange(Map<String, Object> properties) {
+    RateLimiterConfig rateLimiterConfig = getRateLimiterConfig();
+    Map<String, Object> propertiesMap = (Map<String, Object>) properties.get(RL_CONFIG_KEY);
+
+    constructQueryRateLimiterConfigInternal(propertiesMap, rateLimiterConfig);
+  }
+
+  // To be used in initialization
+  @SuppressWarnings({"unchecked"})
+  private static RateLimiterConfig constructQueryRateLimiterConfig(SolrZkClient zkClient) {
+    try {
+
+      if (zkClient == null) {
+        return new RateLimiterConfig(SolrRequest.SolrRequestType.QUERY);
+      }
+
+      Map<String, Object> clusterPropsJson = (Map<String, Object>) Utils.fromJSON(zkClient.getData(ZkStateReader.CLUSTER_PROPS, null, new Stat(), true));
+      Map<String, Object> propertiesMap = (Map<String, Object>) clusterPropsJson.get(RL_CONFIG_KEY);
+      RateLimiterConfig rateLimiterConfig = new RateLimiterConfig(SolrRequest.SolrRequestType.QUERY);
+
+      constructQueryRateLimiterConfigInternal(propertiesMap, rateLimiterConfig);
+
+      return rateLimiterConfig;
+    } catch (KeeperException.NoNodeException e) {
+      return new RateLimiterConfig(SolrRequest.SolrRequestType.QUERY);
+    } catch (KeeperException | InterruptedException e) {
+      throw new RuntimeException("Error reading cluster property", SolrZkClient.checkInterrupted(e));
+    }
   }
 
-  protected static RequestRateLimiter.RateLimiterConfig constructQueryRateLimiterConfig(FilterConfig filterConfig) {
-    RequestRateLimiter.RateLimiterConfig queryRateLimiterConfig = new RequestRateLimiter.RateLimiterConfig();
+  private static void constructQueryRateLimiterConfigInternal(Map<String, Object> propertiesMap, RateLimiterConfig rateLimiterConfig) {
```

Comment:
Why are you not using a POJO and jackson deserialization, like it's done elsewhere? why all these boilerplate code